javatests/arcs/core/analysis/RecipeGraphFixpointIteratorTest.kt

```diff
@@ -0,0 +1,267 @@
+    is RecipeGraph.Node.Particle -> "p:${particle.spec.name}"
+    is RecipeGraph.Node.Handle -> "h:${handle.name}"
+}
+
+/** A simple Fixpoint iterator for the purposes of testing. */
+class TestAnalyzer(
+    val startNames: Set<String>
+) : RecipeGraphFixpointIterator<AbstractSet<String>>() {
+    override fun getInitialValues(graph: RecipeGraph) = graph.nodes.filter {
+        startNames.contains(it.getName())
+    }.associateWith {
+        AbstractSet<String>(setOf())
+    }
+
+    override fun transform(handle: Recipe.Handle, input: AbstractSet<String>) =
+        input.set?.let { AbstractSet<String>(it + "h:${handle.name}") } ?: input
+
+    override fun transform(particle: Recipe.Particle, input: AbstractSet<String>) =
+        input.set?.let { AbstractSet<String>(it + "p:${particle.spec.name}") } ?: input
+
+    override fun transform(
+        handle: Recipe.Handle,
+        particle: Recipe.Particle,
+        spec: HandleConnectionSpec,
+        input: AbstractSet<String>
+    ) = input.set?.let {
+        AbstractSet<String>(it + "h:${handle.name} -> p:${particle.spec.name}")
+    } ?: input
+
+    override fun transform(
+        particle: Recipe.Particle,
+        handle: Recipe.Handle,
+        spec: HandleConnectionSpec,
+        input: AbstractSet<String>
+    ) = input.set?.let {
+        AbstractSet<String>(it + "p:${particle.spec.name} -> h:${handle.name}")
+    } ?: input
+}
+
+@RunWith(JUnit4::class)
+class RecipeGraphFixpointIteratorTest {
+    private val thing = Recipe.Handle("thing", Recipe.Handle.Fate.CREATE, TypeVariable("thing"))
+    private val name = Recipe.Handle("name", Recipe.Handle.Fate.CREATE, TypeVariable("name"))
+    private val readCnxn = HandleConnectionSpec("r", HandleMode.Read, TypeVariable("r"))
+    private val writeCnxn = HandleConnectionSpec("w", HandleMode.Write, TypeVariable("w"))
+    private val readerSpec = ParticleSpec(
+        "Reader",
+        listOf(readCnxn).associateBy { it.name },
+        "ReaderLocation"
+    )
+    private val writerSpec = ParticleSpec(
+        "Writer",
+        listOf(writeCnxn).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val anotherWriterSpec = ParticleSpec(
+        "AnotherWriter",
+        listOf(writeCnxn).associateBy { it.name },
+        "WriterLocation"
+    )
+    private val readerParticle = Recipe.Particle(
+        readerSpec,
+        listOf(Recipe.Particle.HandleConnection(readCnxn, thing))
+    )
+    private val writerParticle = Recipe.Particle(
+        writerSpec,
+        listOf(Recipe.Particle.HandleConnection(writeCnxn, thing))
+    )
+    private val anotherWriterParticle = Recipe.Particle(
+        anotherWriterSpec,
+        listOf(Recipe.Particle.HandleConnection(writeCnxn, thing))
+    )
+
+    @Test
+    fun straightLineFlow() {
+        // [Writer] -> (thing) -> [Reader]
+        val recipe = Recipe(
+            "StraightLine",
+            listOf(thing).associateBy { it.name },
```

Comment:
Not about this PR and a bit of a nit, but still.. I find it a bit annoying that some inputs to constructors (e.g. handles for the Recipe, connection specs as to ParticleSpec above) are required as a Map.

Reason 1) Opportunity for bugs. What if key in the map does not match the object name? We rely that it does, but we don't check it.
Reason 2) It complicates the usage of these objects. E.g. you could probably fit recipe declaration here in one line if map was not required.

If a Recipe (or ParticleSpec) want to have a map keyed by name, they should create the name internally, not expect them to be passed.

Reply:
Yeah, this is certainly getting annoying. I think I based it off some existing code. In any case,  I think it is worth cleaning it up before we get too invested in it.

Reply:
Filed a bug to track. I will take care of this shortly.